tests/benchmarks/random_portfolio_benchmark.cpp

```diff
@@ -0,0 +1,79 @@
+    using namespace std::chrono_literals;
+    std::vector<std::string> assets = {
+        "ABEV3.SAO", "ALPA4.SAO",  "ALSO3.SAO",  "AMAR3.SAO",
+        "AZUL4.SAO", // ASAI3.SAO
+        "B3SA3.SAO", "BBAS3.SAO",  "BBDC3.SAO",  "BBDC4.SAO", "BBSE3.SAO",
+        "BEEF3.SAO", "BPAC11.SAO", "BRAP4.SAO",  "BRDT3.SAO", "BRFS3.SAO",
+        "BRKM5.SAO", "BRML3.SAO",  "BTOW3.SAO",  "CCRO3.SAO", "CESP6.SAO",
+        "CIEL3.SAO", "CMIG4.SAO",  "COGN3.SAO",  "CPFE3.SAO", "CPLE6.SAO",
+        "CRFB3.SAO", "CSAN3.SAO",  "CSMG3.SAO",  "CSNA3.SAO", "CVCB3.SAO",
+        "CYRE3.SAO", "DTEX3.SAO",  "ECOR3.SAO",  "EGIE3.SAO", "ELET3.SAO",
+        "ELET6.SAO", "EMBR3.SAO",  "ENBR3.SAO",  "ENEV3.SAO", "ENGI11.SAO",
+        "EQTL3.SAO", "EZTC3.SAO",  "FLRY3.SAO",  "GGBR4.SAO", "GNDI3.SAO",
+        "GOAU4.SAO", "GOLL4.SAO",  "HAPV3.SAO",  "HGTX3.SAO", "HYPE3.SAO",
+        "IGTA3.SAO", "IRBR3.SAO",  "ITSA4.SAO",  "ITUB4.SAO", "JBSS3.SAO",
+        "JHSF3.SAO", "KLBN11.SAO", "LAME3.SAO",  "LAME4.SAO", "LCAM3.SAO",
+        "LIGT3.SAO", "LINX3.SAO",  "LREN3.SAO",  "MDIA3.SAO", //"LWSA3.SAO"
+        "MEAL3.SAO", "MGLU3.SAO",  "MOVI3.SAO",  "MRFG3.SAO", "MRVE3.SAO",
+        "MULT3.SAO", "NEOE3.SAO",  "NTCO3.SAO",  "PETR3.SAO", "PETR4.SAO",
+        "PRIO3.SAO", //"PCAR3.SAO"
+        "PSSA3.SAO", "QUAL3.SAO",  "RADL3.SAO",  "RAIL3.SAO", "RAPT4.SAO",
+        "RENT3.SAO", "SANB11.SAO", "SAPR11.SAO", "SBSP3.SAO", "SULA11.SAO",
+        "SUZB3.SAO", "TAEE11.SAO", "TIMS3.SAO",  "TOTS3.SAO", "TRPL4.SAO",
+        "UGPA3.SAO", "USIM5.SAO",  "VALE3.SAO",  "VIVT3.SAO", "VVAR3.SAO",
+        "WEGE3.SAO", "YDUQ3.SAO"};
+    portfolio::minute_point mp_start =
+        date::sys_days{2018_y / 01 / 01} + 10h + 0min;
+    portfolio::minute_point mp_end =
+        date::sys_days{2020_y / 12 / 31} + 18h + 0min;
+    portfolio::minute_point start_interval =
+        date::sys_days{2020_y / 01 / 06} + 10h + 0min;
+    portfolio::minute_point end_interval =
+        date::sys_days{2020_y / 01 / 10} + 18h + 0min;
+    portfolio::interval_points interval =
+        std::make_pair(start_interval, end_interval);
+    portfolio::mock_data_feed mock_df;
+    // portfolio::alphavantage_data_feed af("demo", true);
+    portfolio::market_data md(assets, mock_df, mp_start, mp_end,
+                              portfolio::timeframe::weekly, interval, 12);
+    //    portfolio::market_data md(assets, af, mp_start, mp_end,
+    //                              portfolio::timeframe::weekly, interval, 12);
+    portfolio::portfolio port(md);
+    //    random_search<portfolio::market_data, portfolio::portfolio>(
+    //        assets, mock_df, mp_start, mp_end, portfolio::timeframe::daily,
+    //        interval, 30);
+    portfolio::random_search random(md);
+    for (auto _ : state) {
+        random.run(state.range(0));
+        benchmark::DoNotOptimize(random);
+    }
+}
+
+// BENCHMARK(create_vector)->Range(2,2000);
+BENCHMARK(generate_random_portfolio)->Range(2, 2000);
```

Comment:
Veja que o resultado de `random.run(4);` é o mesmo que o resultado de `random.run(2);random.run(2);` e assim por diante. Nestes casos, se você já tem o resultado de `random.run(2);` está desperdiçando (*muito*) recurso ao fazer o `random.run(4);` para descobrir o resultado para `4`.

Neste tipo de experimento, você pode apenas rodar `random.run(1);` várias vezes e sempre guardar o resultado no fim vector. Isso vai dar uma linha com o resultado para cada `n`, que pode plotar com o matplot++. Se quiser, pode refazer tudo várias vezes e ter várias linhas, que podem ser plotadas ou que pode tirar uma média. 

Assim, para cada linha, você vai chamaria o `&random::run` apenas `n` vezes, enquanto do modo como está fazendo, chamaria `&random::run` 1+2+3+4+5+...+n=n*(n+1)/2 vezes. Bem na prática, suponha que `&random::run` demora 1 segundo e queremos testar até n=300. O experimento teria 300/60=5 minutos do primeiro modo enquanto demoraria (300*(300+1))/2/60=752.5 minutos (12 horas e meia) do modo como está fazendo. Em alguns casos, dependendo de `n`, a diferença pode ser muito maior.

Talvez isso é algo que possa fazer até sem o `benchmark` mesmo. Bastaria um `for` colocando os valores no vector. De tempos em tempos você cria um novo plot para poder acompanhar como está o experimento. Só não crie plots demais (todo segundo, por exemplo) porque esses plots também tomam tempo. O objetivo aqui é só te permitir identificar algum problema e parar o experimento caso algo esteja estranho. Não vale a pena atualizar o plot demais pois tomaria tempo do experimento sem ajudar na identificação de erros.

Reply:
Fiz a correção conforme orientação.